feat(fraud-detection): Listen for DataFlow fraud detection events

shane-tomlinson opened the pull request:
Hook up to GCP PubSub to listen for fraud detection events
from the DataFlow fraud detection pipeline.

fixes #311

shane-tomlinson commented:
Pulling WIP off of this. I set up a new project in the GCP console (stomlinson-fxa-dataflow-dev) and a new topic (dataflow), and a new subscription (fxa). I sent a message, and it came through:

<img width="1500" alt="screenshot 2019-02-27 at 12 11 30" src="https://user-images.githubusercontent.com/848085/53489384-f7de1200-3a88-11e9-8b01-2e0cf2902ba2.png">

Some questions:

1. Can we expect a project to be available? If not, not sure how to handle this.
2. Can we expect a topic/subscription to be available? If not, should we create one?
3. If we *can* expect a project/topic/subscription, should we fail hard if they aren't?

@mozilla/fxa-devs - r?

shane-tomlinson commented:
cc @ameihm0912

philbooth commented on lib/server.js:
```diff
@@ -16,12 +16,17 @@ var P = require('bluebird')
 
 module.exports = function createServer(config, log) {
 
+  if (config.dataflow.enabled) {
```
I notice we test this flag twice. You could just make this condiation a nop internally in `lib/dataflow`, then this code wouldn't have to know anything about the internal config of that module.

shane-tomlinson replied:
I like this suggestion.

shane-tomlinson replied:
Made that update.

philbooth commented on test/local/dataflow_tests.js:
```diff
@@ -0,0 +1,162 @@
+let subscription
+
+test(
+  'shared setup', async (t) => {
```
How come `async`? I don't see any `await` or promise-related stuff here or in the module under test...

shane-tomlinson replied:
Because of tap. If the function returns a promise, then tap does not require you to call t.end. However if you do not return a promise, you have to call t.end. It seemed cleaner this way.

philbooth replied:
Ah, makes sense, thanks.

shane-tomlinson commented:
> * Can we expect a project to be available? If not, not sure how to handle this.

I have added logging, if this happens an error is logged but it doesn't prevent the server from starting.

> * Can we expect a topic/subscription to be available? If not, should we create one?

Like above, I have added logging, if this happens an error is logged but it doesn't prevent the server from starting.

> * If we _can_ expect a project/topic/subscription, should we fail hard if they aren't?

I took the approach for testing, no.

> * Can subscriptions/projects ever go away unexpectedly? I'm wondering how the customs server should react if they can.

I tried this, by both deleting subscriptions and topics while the content server is running. No error is displayed by the customs server, I haven't dug in far to determine whether the GCP library surfaces this condition.

ameihm0912 commented:
@shane-tomlinson 

> 1. Can we expect a project to be available? If not, not sure how to handle this.

This will already exist yeah, we have a project setup for production and nonprod use right now for this

> 2. Can we expect a topic/subscription to be available? If not, should we create one?

A topic will already exist. For the subscription, believe we have a couple options. If each instance customs server creates a subscription, they will all get a copy of every message published to the topic. If a subscription is created, and all customs server instances use that single subscription, they will load balance across (each customs instance will get a subset of the published messages).

> 3. If we _can_ expect a project/topic/subscription, should we fail hard if they aren't?

I am guessing here we would want to just log this as a warning and continue, but this scenario should result in an immediate escalation.

> 4. Can subscriptions/projects ever go away unexpectedly?

Projects will not unless they are deleted; likewise with the topics. The subscriptions by default expire after 31 days of inactivity (e.g., no active connections to one). The subscriptions I believe can be configured to persist however.

> 5. What's going to happen as soon as we enable this in prod, are we going to get a flood of events?

Messages sent to a topic with no subscriptions are just dropped. So in the case of enabling it, if it's a new subscription it would just start with any events published after that subscription was created.

> 6. Do we have to worry about rate limiting?

Pubsub will rate limit messages based on how quickly the subscriber is ack'ing them, so there shouldn't need to be any rate limiting in customs required. If it cannot process messages fast enough however it will result in a backlog, but given the expected message rate I don't see this being an issue.

@jbuck feel free to correct me if I'm off anywhere here

jbuck commented:
> > 1. Can we expect a topic/subscription to be available? If not, should we create one?
> 
> A topic will already exist. For the subscription, believe we have a couple options. If each instance customs server creates a subscription, they will all get a copy of every message published to the topic. If a subscription is created, and all customs server instances use that single subscription, they will load balance across (each customs instance will get a subset of the published messages).

I think it makes sense to assume that a subscription exists, and we'll configure the customs server to point at it so that messages are load balanced across all instances of the customs server

> > 2. If we _can_ expect a project/topic/subscription, should we fail hard if they aren't?
> 
> I am guessing here we would want to just log this as a warning and continue, but this scenario should result in an immediate escalation.

Agreed. Monitoring on subscription queue length/age should handle the case where the customs server is not consuming messages

> > 4. Can subscriptions/projects ever go away unexpectedly?
> 
> Projects will not unless they are deleted; likewise with the topics. The subscriptions by default expire after 31 days of inactivity (e.g., no active connections to one). The subscriptions I believe can be configured to persist however.

You could have API failures, but I think it'd be unlikely to actually be deleted in regular use

> @jbuck feel free to correct me if I'm off anywhere here

I think we're all in agreement here
